### ISSUE TITLE
Methods to measure the latency of Torch and FlashAttn

### ISSUE THREAD

**Issue by YJHMITWEB** [Dao-AILab/flash-attention#376]
Hi, thanks for your great work recently. 
In exploring Flash attention, I'd curious about how you measure the throughput (or latency) of this `softmax(Q*K^T)*V` operation on PyTorch. I am aware that in FlashAttention, this is done by calling `flash_attn_func`. I use the following code to measure the time elapsed:
```
b          = 1
q_seqlen   = 2048
kv_seqlen  = 2048
nheads     = 64
headdim    = 128
dropout    = 0.0
softmax_sc = None
causal     = False

q = torch.randn((b, q_seqlen, nheads, headdim), dtype=torch.float16).cuda()
k = torch.randn((b, kv_seqlen, nheads, headdim), dtype=torch.float16).cuda()
v = torch.randn((b, kv_seqlen, nheads, headdim), dtype=torch.float16).cuda()

torch.cuda.synchronize()
start_time = time.time()
flash_attn_func(q, k, v)
torch.cuda.synchronize()
end_time = time.time()
fa_time = end_time - start_time

def torch_mha(q, k, v):
    scores = torch.einsum('bthd,bshd->bhts', q / math.sqrt(q.shape[-1]), k)
    attention = torch.softmax(scores, dim=-1)
    output = torch.einsum('bhts,bshd->bthd', attention, v)
    
torch.cuda.synchronize()
start_time = time.time()
torch_mha(q, k, v)
torch.cuda.synchronize()
end_time = time.time()
torch_time = end_time - start_time

print("Torch MHA: ", torch_time, ", Flash Attn: ", fa_time)
```
On A100 GPU, the result is like `Torch MHA:  0.49175071716308594 , Flash Attn:  0.005777835845947266`, which shows about 100x speedup, which is even faster than stated in the paper.

I am wondering if this is the right way to measure it.

**Comment by tridao** [comment]
You probably want to use Pytorch's [benchmark](https://pytorch.org/tutorials/recipes/recipes/benchmark.html) util.
Above you're just running each function once. That might not be enough to get reliable benchmark numbers. Pytorch benchmark util will take care of running several times including warming up.

**Comment by YJHMITWEB** [comment]
Thanks, I am using pytorch benchmark and the result looks much more like in the paper.
